### ISSUE TITLE
fix: test_training_loop DataLoader cannot be converted to PythonObject in run_epoch call

### ISSUE THREAD

**Issue by mvillmow** [HomericIntelligence/ProjectOdyssey#4522]
## Problem

Two training loop test files fail with --Werror:

\`\`\`
tests/shared/training/test_training_loop.mojo:118:33: error: invalid call to 'run_epoch': value passed to 'data_loader' cannot be converted from 'DataLoader' to 'PythonObject'
tests/shared/training/test_training_loop.mojo:129:34: error: invalid call to 'run_epoch': value passed to 'data_loader' cannot be converted from 'DataLoader' to 'PythonObject'
\`\`\`

Same error in \`tests/shared/training/test_training_loop_part1.mojo:118\`.

## Root Cause

The \`run_epoch\` function signature expects \`PythonObject\` for \`data_loader\` parameter, but
the test is passing a Mojo \`DataLoader\` type directly.

## Fix Options

1. Add a \`run_epoch\` overload that accepts the Mojo \`DataLoader\` type
2. Update the test to wrap the DataLoader in a PythonObject
3. Update \`run_epoch\` to accept \`DataLoader\` directly

## Affected Files

- \`tests/shared/training/test_training_loop.mojo\`
- \`tests/shared/training/test_training_loop_part1.mojo\`

## Context

Found during comprehensive --Werror compilation audit (PR #4512, branch `batch/low-complexity-fixes`).

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have all the context needed. Here's the implementation plan:

---

## Implementation Plan: Issue #4522 — Fix `run_epoch` DataLoader Type Mismatch

---

### 1. Objective

Add a `DataLoader` overload to `run_epoch` so that tests can pass a native Mojo `DataLoader` directly, eliminating the `cannot be converted from 'DataLoader' to 'PythonObject'` compile errors in `test_training_loop.mojo:118,129` and `test_training_loop_part1.mojo:118`.

---

### 2. Approach

**Add a `DataLoader` overload that delegates to actual batch iteration logic.**

The existing `PythonObject` overload is a stub (it does nothing, returns `0.0`). The `DataLoader` struct is fully functional with `has_next()` / `next()`. The fix is to add a second `run_epoch` overload that accepts the Mojo `DataLoader` and actually iterates it — this also makes the implementation correct, not just compilable.

Key decisions:
- **Add overload, don't change the `PythonObject` signature** — the original is documented as pending Track 4 (#3076); leave it intact
- **Implement real batch iteration in the new overload** — the `DataLoader` supports it, so implement it properly (avoids a trivial pass-through that would still fail assertions in `test_training_loop.mojo`)
- **Minimal scope** — only touch `shared/training/__init__.mojo`; no test changes needed

---

### 3. Files to Create

None — no new files required.

---

### 4. Files to Modify

**`shared/training/__init__.mojo`**

- Add a second `run_epoch` method overload that accepts `DataLoader` (from `shared.training.trainer_interface`):

  ```mojo
  fn run_epoch(mut self, mut data_loader: DataLoader) raises -> Float32:
      data_loader.reset()
      var total_loss = Float64(0.0)
      var num_batches = 0
      while data_loader.has_next():
          var batch = data_loader.next()
          var loss = self.train_step(batch.data, batch.labels)
          total_loss += Float64(loss)
          num_batches += 1
      if num_batches > 0:
          return Float32(total_loss / Float64(num_batches))
      return Float32(0.0)
  ```

- Ensure `DataLoader` and `DataBatch` are imported at the top of the file (check if already imported via `trainer_interface`).

---

### 5. Implementation Order

1. **Read** `shared/training/__init__.mojo` fully to confirm current imports and `train_step` signature.
2. **Read** `shared/training/trainer_interface.mojo` to confirm `DataLoader`/`DataBatch` are already importable from within `__init__.mojo` or need an explicit import.
3. **Edit** `shared/training/__init__.mojo`: add the `DataLoader` overload of `run_epoch` after the existing `PythonObject` overload.
4. **Verify** by pushing to CI (local GLIBC constraints prevent running `mojo test` locally).

---

### 6. Verification

- **CI must pass** the two affected test files:
  - `tests/shared/training/test_training_loop.mojo` — lines 118, 129 call `run_epoch(data_loader)` where `data_loader: DataLoader`
  - `tests/shared/training/test_training_loop_part1.mojo` — line 118 same call
- **Assertions to satisfy**:
  - `test_training_loop.mojo`: `assert_greater(Float64(avg_loss), Float64(0.0))` — the new overload must actually compute loss (not return `0.0`), so real batch iteration is required
  - `test_training_loop_part1.mojo`: `assert_greater(Float64(avg_loss), Float64(-0.001))` — passes even with `0.0` return, but real iteration is still preferable
- **No regression** on any existing tests that use the `PythonObject` overload

---

### 7. Skills Used

**During planning:**
- `docs-search` / `graph-query` — used via the `Explore` subagent to locate `run_epoch`, `DataLoader`, and test files

**From Prior Learnings / Team Knowledge Base:**
- `mojo-float-literal-overload-fix` — direct pattern: add a typed overload that delegates/implements the logic, matching the Float32→Float64 delegation pattern
- `fix-ci-mojo-test-failure` — systematic workflow: investigate → read implementation → apply minimal fix; verify via CI not local `mojo test`
- `mojo-setitem-test-pattern` — pattern for testing multiple overloads with different types
- `mojo-trait-conformance-fix` — consulted for any `DataLoader` trait conformance requirements (Copyable/Movable already satisfied)

---
*Generated by Claude Code Planner*
